File: src/p0020/golang/solution.go
func isValid(s string) bool {
	var stk []byte
	for i := 0; i < len(s); i++ {
		if len(stk) == 0 {
			stk = append(stk, s[i])
		} else {
			if isMatch(stk[len(stk)-1], s[i]) {
				stk = stk[:len(stk)-1]
			} else {
				stk = append(stk, s[i])
			}
		}
	}
	return len(stk) == 0
}

func isMatch(p byte, q byte) bool {
	return p == '(' && q == ')' ||
		p == '[' && q == ']' ||
		p == '{' && q == '}'
}

File: src/p0020/rust/solution.rs
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let is_match = |p, q| {
            p == '(' && q == ')' ||
            p == '[' && q == ']' ||
            p == '{' && q == '}'
        };
        let mut stk = Vec::new();
        for x in s.chars() {
            if stk.is_empty() {
                stk.push(x)
            } else {
                if is_match(*stk.last().unwrap(), x) {
                    stk.pop();
                } else {
                    stk.push(x);
                }
            }
        }
        return stk.is_empty();
    }
}
